### PR TITLE
Core: Fix single instance windows

### LINKED ISSUE #15433: Core: Single Instance flag does not work on Windows
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem description

Having a FreeCAD instance open and running `FreeCAD.exe --single-instance <freecad file>` does not open the file.

### Full version info

```shell
OS: Arch Linux
Word size of FreeCAD: 64-bit
Version: 0.22.0dev.38187 (Git)
Build type: DEBUG
Branch: main
Hash: 2a460a081303eddc645d35db1e82d7c3e33cce04
Python 3.12.4, Qt 6.7.2, Coin 4.0.2, Vtk , OCC 7.7.2
Locale: English/United States (en_US)
```

### Subproject(s) affected?

Core

### Anything else?

I believe this is a known, potentially old problem that has not been reported, because in `tools/build/WindowsInstaller/setup/configure.nsh`, it says this: `WriteRegStr SHCTX "Software\Classes\Applications\{BIN_FREECAD}\shell\open\command" "" '"$INSTDIR\${APP_RUN}" "%1"'` (without `--single-instance`).
In contrast, in `src/XDGData/org.freecad.FreeCAD.Desktop` it says this: `Exec=FreeCAD - --single-instance %F` (with `--single-instance`). 

The `--single-instance` flag makes sure that FreeCAD finds an already running FreeCAD instance to open the files there.

The effect is that if you click on a file in Windows it will always open a new instance of FreeCAD, whereas on Linux it will open an existing instance of FreeCAD (if applicable) to open the file there. I assume this would also be the preferred behavior on Windows.

I believe the problem lies with connection type between a server and client FreeCAD, possibly having to do with line endings that prevent from flushing. I will submit a PR for this.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by pieterhijma**
Closes #15433.

Note that the last commit adds the `--single-instance` to the Windows installer but unfortunately I'm not in a position to test this.

The fix is a switch from using QDataStream/QByteArray to QTextStream/QString combined with explicitly setting the encoding for transfer, adding newlines per message and reading lines on the other side of the socket. Additionally, I switched from explicitly waiting for a read to react on the `readyRead` signal because the first attempt on trying to read always fails on Windows after which a second one succeeds. I cannot explain this, but reacting to a `readyRead` signal is in my opinion the best way to handle this small issue.

The PR doesn't include tests because two FreeCAD instances are needed to test the behavior. 

To test, first start FreeCAD:

```
FreeCAD.exe
```

This should start the FreeCAD GUI. Then executing the command below should open `cube.FCStd` in the previous FreeCAD instance.

```
FreeCAD.exe --single-instance cube.FCStd
```

If the installer has run, clicking on FreeCAD files should open the files in the first (and hopefully) only FreeCAD instance, providing Windows users the same behavior that Linux users enjoy.

### CODE REVIEW

**Comment by pieterhijma** [comment]
@PaddleStroke, could you perhaps test the installer (or edit the Windows register to include `--single-instance` in the shell commands specified in the last commit)?

File: src/Gui/GuiApplication.cpp [hunk 1]
```diff
@@ -237,57 +237,83 @@ bool GUISingleApplication::isRunning() const
...
+#if QT_VERSION <= QT_VERSION_CHECK(5, 15, 0)
+    ts << message << endl;
+#else
+    ts << message << Qt::endl;
+#endif
+
+    bool bytesWritten = socket.waitForBytesWritten(timeout);
+    socket.close();
+    if (socket.state() == QLocalSocket::UnconnectedState ||
+        socket.waitForDisconnected()) {
```
Comment (PaddleStroke) [inline src/Gui/GuiApplication.cpp:276]: Lint warning, the if is empty.

Comment (pieterhijma) [reply src/Gui/GuiApplication.cpp:276]: Ah yes, thanks, I missed this one when removing debugging code.

File: src/Gui/Application.cpp [hunk 2]
```diff
@@ -2001,10 +2001,9 @@ bool onlySingleInstance(GUISingleApplication& mainApp)
                 fn = QDir::cleanPath(fn);
             }
 
-            QByteArray msg = fn.toUtf8();
-            msg.prepend("OpenFile:");
-            if (!mainApp.sendMessage(msg)) {
-                qWarning("Failed to send message to server");
+            fn.prepend(QString::fromStdString("OpenFile:"));
```
Comment (chennes) [inline src/Gui/Application.cpp:2004]: No reason to roundtrip through `std::string` here, right? You could use `QLatin1StringView` or one of the other related classes.

**Comment by chennes** [comment]
I can confirm that this fix works, tested with:
```
OS: Windows 11 build 22631
Word size of FreeCAD: 64-bit
Version: 0.22.0dev.38260 +7 (Git)
Build type: Release
Branch: pr15504
Hash: 2dfe75bf51557f93b5acfdfbd178742af33a3477
Python 3.12.4, Qt 6.7.2, Coin 4.0.2, Vtk 9.3.0, OCC 7.8.0
Locale: English/United States (en_US)
```

File: src/Gui/Application.cpp (hunk 2 above)
Comment (pieterhijma) [reply src/Gui/Application.cpp]: No, indeed, I used `QLatin1String` because it works for both Qt 5 and 6.

**Comment by chennes** [comment]
Retested with Qt5, works as expected in Windows:
```
OS: Windows 11 build 22631
Word size of FreeCAD: 64-bit
Version: 0.22.0dev.38318 +3 (Git)
Build type: Release
Python 3.8.10, Qt 5.15.2, Coin 4.0.1, Vtk 8.2.0, OCC 7.6.3
Locale: English/United States (en_US)
```

**Comment by chennes** [comment]
Can I get a Linux user to verify that there are no unforeseen issues on Linux with this PR? @adrianinsaval? @kadet1090?

**Pull request merged**